--- a/miniupnpd.c
+++ b/miniupnpd.c
@@ -1195,6 +1195,7 @@ static void update_disable_port_forwarding(void)
 		if (!disable_port_forwarding && reserved) {
 			if (GETFLAG(ALLOWPRIVATEIPV4MASK)) {
 				syslog(LOG_WARNING, "IGNORED : Reserved / private IP address %s on ext interface %s", if_addr, ext_if_name);
+				disable_port_forwarding = 0;
 			} else {
 				syslog(LOG_WARNING, "Reserved / private IP address %s on ext interface %s: Port forwarding is impossible", if_addr, ext_if_name);
 				syslog(LOG_INFO, "You are probably behind NAT, enable option ext_perform_stun=yes to detect public IP address");
@@ -1208,6 +1209,9 @@ static void update_disable_port_forwarding(void)
 		} else if (disable_port_forwarding && !reserved) {
 			syslog(LOG_INFO, "Public IP address %s on ext interface %s: Port forwarding is enabled", if_addr, ext_if_name);
 			disable_port_forwarding = 0;
+		} else if (disable_port_forwarding && reserved && GETFLAG(ALLOWPRIVATEIPV4MASK)) {
+			syslog(LOG_INFO, "Public IP address %s on ext interface %s: Port forwarding is enabled", if_addr, ext_if_name);
+			disable_port_forwarding = 0;
 		}
 	}
 }
